linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -5434,16 +5590,58 @@ private[optimizer] object OptimizerCore {
         }
       }
 
-      new MethodAttributes(inlineable, shouldInline, isForwarder)
+      val jsDynImportInlineTarget = body match {
+        case MaybeUnbox(SelectJSNativeMember(className, MethodIdent(member)), _) =>
+          Some(ImportTarget.Member(className, member))
+
+        case MaybeUnbox(JSFunctionApply(SelectJSNativeMember(className,
+            MethodIdent(member)), args), _) if args.forall(isSmallTree(_))=>
+          Some(ImportTarget.Member(className, member))
+
+        case MaybeUnbox(LoadJSModule(className), _) =>
+          Some(ImportTarget.Class(className))
+
+        case MaybeUnbox(JSSelect(LoadJSModule(className), arg), _) if isSmallTree(arg) =>
+          Some(ImportTarget.Class(className))
+
+        case MaybeUnbox(JSMethodApply(LoadJSModule(className), method, args), _)
+            if isSmallTree(method) && args.forall(isSmallTree(_)) =>
+          Some(ImportTarget.Class(className))
+
+        case JSNew(LoadJSConstructor(className), args) if args.forall(isSmallTree(_)) =>
+          Some(ImportTarget.Class(className))
+
+        case _ =>
+          None
+      }
+
+      val jsDynImportThunkFor = body match {
+        case Apply(_, New(clazz, _, _), MethodIdent(target), _) if clazz == enclosingClassName =>
+          Some(target)
```

Comment:
IIUC, when this case applies, we will ignore the constructor of `clazz` and its superclasses. Either we need to check that they are side-effect-free somehow, or (probably easier), we should preserve the `New`, and then `transformStat` it somewhere down the line (which, when the class is inlineable and its constructor is indeed side-effect-free, will eventually collapse to `Skip()`).

Reply:
I'm not sure I understand. This is just a heuristics of whether we should even *try* to inline the `ApplyDynamicImport`. The full inlining machinery is used to actually inline (or not).